hircluster.c

```diff
@@ -3741,34 +3738,42 @@ void clusterNodesReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
     }
 }
 
-/* Select a node. Find a connected node during 3 random picks,
- * otherwise pick a node where a connection establishment has not
- * been attempted within a throttle time. */
+/* Select a node.
+ * Primarily selects a connected node found close to a randomly picked index of
+ * all known nodes. The random index should give a more even distribution of
+ * selected nodes. If no connected node is found while iterating to this index
+ * any node where a connect has not been attempted within throttle-time is selected.
+ * All nodes are checked until a connected or accepted node is found.
+ */
 static redisClusterNode *selectNode(dict *nodes) {
-    redisClusterNode *node;
-    dictEntry *de;
-
-    /* Perform 3 attempts to find a random connected node */
-    for (int i = 0; i < 3; i++) {
-        if ((de = dictGetRandomKey(nodes)) != NULL) {
-            node = dictGetEntryVal(de);
-            if (node->acon != NULL && node->acon->err == 0 &&
-                node->acon->c.flags & REDIS_CONNECTED) {
-                return node;
-            }
-        }
-    }
+    redisClusterNode *node, *connected = NULL, *accepted = NULL;
+    dictIterator di;
+    dictInitIterator(&di, nodes);
 
-    /* Find any node which has not been attempted within throttle limits */
     int64_t throttleLimit = hi_usec_now() - SLOTMAP_UPDATE_THROTTLE_USEC;
+    unsigned long currentIndex = 0;
+    unsigned long checkIndex = random() % dictSize(nodes);
 
-    dictIterator di;
-    dictInitIterator(&di, nodes);
+    dictEntry *de;
     while ((de = dictNext(&di)) != NULL) {
         node = dictGetEntryVal(de);
+        /* Keep any connected node */
+        if (node->acon != NULL && node->acon->err == 0 &&
+            node->acon->c.flags & REDIS_CONNECTED) {
+            connected = node;
+        }
+        /* Keep any node which has not been attempted within throttle limits */
         if (node->lastConnectionAttempt < throttleLimit) {
-            return node;
+            accepted = node;
+        }
+        /* Return a connected or accepted node when chosen index is reached */
+        if (currentIndex >= checkIndex) {
+            if (connected != NULL)
+                return connected;
+            if (accepted != NULL)
+                return accepted;
         }
```

Comment:
Corner case: `checkIndex == 0` and the first node is not connected. In this case, we would pick the first node even if it's not connected.

To avoid it, we can run a few more steps to look for a connected node, before we return accepted.

```suggestion
        /* Return a connected or accepted node when chosen index is reached */
        if (currentIndex >= checkIndex && connected != NULL)
            return connected;
        else if (currentIndex >= checkIndex + 2 && accepted != NULL)
            return accepted;
```

Reply:
Hum, this would depend on the amount of nodes
If we only have one node, and that one is not connected but accepted, this would return NULL/none then?
Same with two nodes.

Reply:
Yes, should be safe. The `while ((de = dictNext(&di)) != NULL) {}` breaks the loop when there are no more nodes in the list.

Reply:
Easy fixable, added a new fixup commit